packages/cloud-config/oem/08_boot_assessment.yaml

```diff
@@ -0,0 +1,111 @@
+# The boot assessment works in this way:
+# - After we upgrade, we set a grub env sentinel variable indicating that we did run an upgrade
+# - At the first boot, if we did an upgrade, we set another sentinel variable, which indicates we are trying to boot
+# - If we boot fine, we remove both sentinels
+# - If we get back again at the GRUB menu, a failure must have occurred and we select the fallback entry, creating also
+#   sentinels files and a specific cmdline option indicating we failed booting after an upgrade
+
+name: "Boot assessment"
+stages:
+    boot.before:
+    # Remove any grub sentinel and add
+    # System sentinel to /run to indicate if we are booting after a failed upgrade.
+    # This needs to run only on the active partition.
+    - name: "Remove upgrade sentinels"
+      if: |
+          cat /proc/cmdline | grep -q "active.img"
+      commands:
+      - |
+          sudo mount -o rw,remount /run/initramfs/cos-state
+          grub2-editenv /run/initramfs/cos-state/boot_assessment set osupgrade=
+          grub2-editenv /run/initramfs/cos-state/boot_assessment set tentative=
+          sudo mount -o ro,remount /run/initramfs/cos-state
+    - name: "Create upgrade failure sentinel if necessary"
+      if: |
+          cat /proc/cmdline | grep -q "upgrade_failure"
+      files:
+      - path: /run/cos/upgrade_failure
+        content: "1"
+        permissions: 0600
+        owner: 0
+        group: 0
+    after-install:
+    # After install, reset, and upgrade, we install additional GRUB configuration for boot assessment into COS_STATE.
+
+    - &statemount
+      name: "Mount state"
+      commands:
+      - |
+          STATEDIR=/tmp/mnt/STATE
+          STATE=$(blkid -L COS_STATE || true)
+          mkdir -p $STATEDIR || true
+          mount ${STATE} $STATEDIR
+    # Here we hook the boot assessment configuration to 'grubcustom'
+    # we do that selectively in order to just "append" eventual other configuration provided.
+    # XXX: maybe we should just write to /grubcustom and override any other custom grub?
+    - &customhook
+      name: "Hook boot assessment grub configuration"
+      if: |
+          [ ! -f "/tmp/mnt/STATE/grubcustom" ] || ( cat /tmp/mnt/STATE/grubcustom | grep -v "grub_boot_assessment" )
+      commands:
+      - |
+        cat << EOF >> /tmp/mnt/STATE/grubcustom
+        set bootfile="/grub_boot_assessment"
+        search --no-floppy --file --set=bootfile_loc "${bootfile}"
+        if [ "${bootfile_loc}" ]; then
+           source "(${bootfile_loc})${bootfile}"
+        EOF
+    # Overrides the active cmdline by adding "rd.emergency=reboot", "rd.shell=0" and "panic=5"
```

Comment:
@davidcassany in my tests, I managed to get the initramfs stuck by removing "usr', "sbin", "lib" from the image. It stayed to initramfs stuck after I could see it mounted the images (the debug messages where displaying ext4 mount info text blobs)

I then tried to reduce the dracut timeout to something smaller without success (with `rd.timeout`), so it seems in certain cases the initramfs stays still in the foreground trying to do something (I couldn't check what exactly). However, didn't seemed related to any dracutinitqueue timeout afaict, but any pointer might be appreciated.

The thing is, If we have an option to let dracut start in a time range would be good as well. so in any case we can restart into fallback even after a long time passed.

Anyway, this mechanism seems to be flexible enough, as a reboot would automatically select fallback in that case as well. it is a matter on how dracut is configured at that point

Reply:
mm we should analyze this specific case. IIRC `rd.timeout` is only honored by the `initqueue` so if your problem was before/beyond the initqueue then it makes sense rd.timeout was simply ignored.

Note the boot is systemd based, so it can happen some early boot services are in starting state forever, like mountpoints without a timeout or with an insane timeout. So it is likely we need to work per use case basis... this is hard, since there things that are not directly configured by us. 

We could try to analyze the boot procedure and try to find services that could stuck forever, note this has potential to be distro specific. 

Dracut bootup procedure is documented here https://man7.org/linux/man-pages/man7/dracut.bootup.7.html

Reply:
I've been able to trace back my attempt and the specific case is triggered by removing the following folders from the image and trying it to boot: "bin", "usr", "sbin", "lib".

That triggers a boot stuck which doesn't seem to honor `rd.timeout`. Note I can see the mounting messages of the images, so we get past to the point where images are loaded. Interestingly, doesn't seems to trigger a kernel panic.